pyvisfile/xdmf/__init__.py

```diff
@@ -0,0 +1,1119 @@
+    DG = enum.auto()    # Discontinuous Galerkin elements for simplices
+    CG = enum.auto()    # Continuous Galerkin elements for simplices
+    Q = enum.auto()     # Continuous Galerkin elements for quadrilaterals
+    DQ = enum.auto()    # Discontinuous Galerkin elements for quadrilaterals
+    RT = enum.auto()    # Raviart-Thomas elements on triangles
+
+
+class AttributeElementCell(enum.Enum):
+    """Reference element types for ``FiniteElementFunction``."""
+    interval = enum.auto()
+    triangle = enum.auto()
+    tetrahedron = enum.auto()
+    quadrilateral = enum.auto()
+    hexahedron = enum.auto()
+
+# }}}
+
+
+def _attribute_type_from_shape(shape):
+    if len(shape) == 1 or (len(shape) == 2 and shape[1] == 1):
+        return AttributeType.Scalar
+    elif len(shape) == 2 and shape[0] in [2, 3]:
+        return AttributeType.Vector
+    elif len(shape) == 2 and shape[0] in [4, 9]:
+        return AttributeType.Tensor
+    elif len(shape) == 2 and shape[0] == 6:
+        return AttributeType.Tensor6
+    elif len(shape) == 3:
+        return AttributeType.Matrix
+    else:
+        raise ValueError(f"cannot determine attribute type from shape '{shape}'")
+
+
+class AttributeType(enum.Enum):
+    """Rank of the attribute stored on the mesh."""
+    # NOTE: integer ids taken from ``XdmfAttributeType.hpp``
+    Scalar = 200
+    Vector = 201
+    Tensor = 202
+    Matrix = 203
+    Tensor6 = 204
+    GlobalId = 205
+
+
+class AttributeCenter(enum.Enum):
+    """Center of the attribute stored on the mesh."""
+    # NOTE: integer ids taken from ``XdmfAttributeCenter.hpp``
+    Grid = 100
+    Cell = 101
+    Face = 102
+    Edge = 103
+    Node = 104
+    # NOTE: only for `FiniteElementFunction` attributes, which are not supported
+    Other = 105
+
+
+class Attribute(XdmfElement):
+    """
+    .. automethod:: __init__
+    """
+
+    def __init__(self, *,
+            name: Optional[str] = None,
+            atype: AttributeType = AttributeType.Scalar,
+            acenter: AttributeCenter = AttributeCenter.Node,
+            parent: Optional[Element] = None,
+            ):
+        """
+        :param parent: if provided, *self* is appended to the element.
+        """
+
+        super().__init__(parent, "Attribute", {
+            "Name": name,
+            "Center": acenter.name,
+            "AttributeType": atype.name,
+            })
+
+# }}}
+
+
+# {{{ data item
+
+def _numpy_to_xdmf_number_type(dtype):
+    if dtype.type in (np.int8, np.int16, np.int32, np.int64, np.int):
+        return DataItemNumberType.Int
+    elif dtype.type in (np.uint8, np.uint16, np.uint32, np.uint64, np.uint):
+        return DataItemNumberType.UInt
+    elif dtype.type in (np.float16, np.float32, np.float64, np.float128):
+        return DataItemNumberType.Float
+    else:
+        raise ValueError(f"unsupported dtype: '{dtype}'")
+
+
+def _data_item_format_from_str(text):
+    if not isinstance(text, str):
+        return DataItemFormat.XML
```

Comment:
That feels too loose.

Reply:
Removed this all together and made it just handle strings in 11238ea.